src/lib/marlin_plonk_bindings/stubs/src/tweedle_fp.rs

```diff
@@ -0,0 +1,214 @@
+use crate::bigint_256::{CamlBigint256, CamlBigint256Ptr};
+use algebra::{
+    fields::{Field, FpParameters, PrimeField, SquareRootField},
+    tweedle::fp::{Fp, FpParameters as Fp_params},
+    FftField, One, UniformRand, Zero,
+};
+use ff_fft::{EvaluationDomain, Radix2EvaluationDomain as Domain};
+use num_bigint::BigUint;
+use rand::rngs::StdRng;
+use std::cmp::Ordering::{Equal, Greater, Less};
+
+#[derive(Copy, Clone)]
+pub struct CamlTweedleFp(Fp);
+
+pub type CamlTweedleFpPtr = ocaml::Pointer<CamlTweedleFp>;
+
+extern "C" fn caml_tweedle_fp_compare_raw(x: ocaml::Value, y: ocaml::Value) -> libc::c_int {
+    let x: CamlTweedleFpPtr = ocaml::FromValue::from_value(x);
+    let y: CamlTweedleFpPtr = ocaml::FromValue::from_value(y);
+
+    match x.as_ref().0.cmp(&y.as_ref().0) {
+        Less => -1,
+        Equal => 0,
+        Greater => 1,
+    }
+}
+
+impl From<&CamlTweedleFp> for CamlBigint256 {
+    fn from(x: &CamlTweedleFp) -> CamlBigint256 {
+        CamlBigint256(x.0.into_repr())
+    }
+}
+
+impl From<&CamlBigint256> for CamlTweedleFp {
+    fn from(x: &CamlBigint256) -> CamlTweedleFp {
+        CamlTweedleFp(Fp::from_repr(x.0))
+    }
+}
+
+impl std::fmt::Display for CamlTweedleFp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        CamlBigint256::from(self).fmt(f)
+    }
+}
+
+ocaml::custom!(CamlTweedleFp {
+    compare: caml_tweedle_fp_compare_raw,
+});
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_size_in_bits() -> ocaml::Int {
+    Fp_params::MODULUS_BITS as isize
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_size() -> CamlBigint256 {
+    CamlBigint256(Fp_params::MODULUS)
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_add(x: CamlTweedleFpPtr, y: CamlTweedleFpPtr) -> CamlTweedleFp {
+    CamlTweedleFp(x.as_ref().0 + y.as_ref().0)
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_sub(x: CamlTweedleFpPtr, y: CamlTweedleFpPtr) -> CamlTweedleFp {
+    CamlTweedleFp(x.as_ref().0 - y.as_ref().0)
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_negate(x: CamlTweedleFpPtr) -> CamlTweedleFp {
+    CamlTweedleFp(-x.as_ref().0)
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_mul(x: CamlTweedleFpPtr, y: CamlTweedleFpPtr) -> CamlTweedleFp {
+    CamlTweedleFp(x.as_ref().0 * y.as_ref().0)
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_div(x: CamlTweedleFpPtr, y: CamlTweedleFpPtr) -> CamlTweedleFp {
+    CamlTweedleFp(x.as_ref().0 / y.as_ref().0)
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_inv(x: CamlTweedleFpPtr) -> Option<CamlTweedleFp> {
+    match x.as_ref().0.inverse() {
+        Some(x) => Some(CamlTweedleFp(x)),
+        None => None,
+    }
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_square(x: CamlTweedleFpPtr) -> CamlTweedleFp {
+    CamlTweedleFp(x.as_ref().0.square())
+}
+
+#[ocaml::func]
+pub fn caml_tweedle_fp_is_square(x: CamlTweedleFpPtr) -> bool {
```

Comment:
I noticed some functions take `CamlTweedleFp` and some take `CamlTweedleFpPtr` -- what determines that?

Reply:
`CamlTweedleFp` is the representation,  `CamlTweedleFpPtr` is effectively a `&CamlTweedleFp` where the reference points to the OCaml heap. We can write `CamlTweedleFp`s, but reading them has to go through `CamlTweedlePtr` to handle the fact that the OCaml GC can relocate the data.